### PR TITLE
deprecate azure_germany

### LINKED ISSUE #37653: Deprecate AzureAuthorityHosts entry for German Cloud
For context, see https://github.com/Azure/azure-sdk-for-net/issues/46311. Deprecate the following: https://github.com/Azure/azure-sdk-for-python/blob/a88c14555f6305c5beaa81ab243269eb5a64bc7a/sdk/identity/azure-identity/azure/identity/_constants.py#L18

### PR DESCRIPTION
**Pull request by xiangyan99**
Closes #37653

### CODE REVIEW

**Review by pvaneck — COMMENTED** [review]
Probably misleading to call this "deprecate' when it is being removed here. We can follow a similar approach as [this](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py#L235) if we want to keep it useable but warn users that do actually use it.

**Comment by xiangyan99** [comment]
> Probably misleading to call this "deprecate' when it is being removed here. We can follow a similar approach as [this](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py#L235) if we want to keep it useable but warn users that do actually use it.

Azure Germany was closed in 2021. The code is broken. Why we leave the option to let user choose the wrong one?

**Comment by pvaneck** [comment]
> Azure Germany was closed in 2021. The code is broken. Why we leave the option to let user choose the wrong one?

This is still a public API, and we shouldn't completely remove it without fair warning. People might still be referencing it, and we wouldn't want to suddenly break them with `AttributeErrors` (one example I found searching GitHub [here](https://github.com/prowler-cloud/prowler/blob/master/prowler/providers/azure/lib/regions/regions.py#L27)).

**Comment by xiangyan99** [comment]
> > Azure Germany was closed in 2021. The code is broken. Why we leave the option to let user choose the wrong one?
> 
> This is still a public API, and we shouldn't completely remove it without fair warning. People might still be referencing it, and we wouldn't want to suddenly break them with `AttributeErrors` (one example I found searching GitHub [here](https://github.com/prowler-cloud/prowler/blob/master/prowler/providers/azure/lib/regions/regions.py#L27)).

Is the code still working?

I don't think any code that targets Azure Germany still works.

**Comment by pvaneck** [comment]
> Is the code still working?
> 
> I don't think any code that targets Azure Germany still works.

It's just generating a config, so the code still works. Now, if any request is actually made to the German cloud, then I'm sure it'd fail. But the fact still remains that they are using `AZURE_GERMANY`, and we shouldn't break them (and others) without warning.

**Comment by xiangyan99** [comment]
> > Is the code still working?
> > I don't think any code that targets Azure Germany still works.
> 
> It's just generating a config, so the code still works. Now, if any request is actually made to the German cloud, then I'm sure it'd fail. But the fact still remains that they are using `AZURE_GERMANY`, and we shouldn't break them (and others) without warning.

To make sure we are on the same page, the code is like

cred = DAC(authority_host=AzureAuthorityHosts.AZURE_GERMANY)

The code is broken today. This was a breaking change on the service side. There is no point we give a warning rather than breaking it.

**Comment by scottaddie** [comment]
If we remove it, will this change ship in a major release?

**Comment by xiangyan99** [comment]
> If we remove it, will this change ship in a major release?

The problem is no matter we keep it or not, the code does not work. It is a breaking change on the service side.

I don't think we need a major version bump for this.

**Comment by pvaneck** [comment]
> To make sure we are on the same page, the code is like
> 
> cred = DAC(authority_host=AzureAuthorityHosts.AZURE_GERMANY)
> 
> The code is broken today. This was a breaking change on the service side. There is no point we give a warning rather than breaking it.

Sure, that code won't work, but some user code might access it but not actually use it for auth. Just accessing `AzureAuthorityHosts.AZURE_GERMANY` would now be broken with full removal. We see that there are some cases of people referencing this value, so let's not cause their apps to suddenly break with AttributeErrors (at least not yet without warning).

**Comment by xiangyan99** [comment]
> > To make sure we are on the same page, the code is like
> > cred = DAC(authority_host=AzureAuthorityHosts.AZURE_GERMANY)
> > The code is broken today. This was a breaking change on the service side. There is no point we give a warning rather than breaking it.
> 
> Sure, that code won't work, but some user code might access it but not actually use it for auth. Just accessing `AzureAuthorityHosts.AZURE_GERMANY` would now be broken with full removal. We see that there are some cases of people referencing this value, so let's not cause their apps to suddenly break with AttributeErrors (at least not yet without warning).

If user has explicit `AzureAuthorityHosts.AZURE_GERMANY`, we cannot assume they don't want to use it. :)

If we know for sure the code is wrong, let's "force" customers to fix it.

Sometimes it can be caught by typing tools (static check). It will be much better to crash/not work after the deployment.

**Comment by xiangyan99** [comment]
> > To make sure we are on the same page, the code is like
> > cred = DAC(authority_host=AzureAuthorityHosts.AZURE_GERMANY)
> > The code is broken today. This was a breaking change on the service side. There is no point we give a warning rather than breaking it.
> 
> Sure, that code won't work, but some user code might access it but not actually use it for auth. Just accessing `AzureAuthorityHosts.AZURE_GERMANY` would now be broken with full removal. We see that there are some cases of people referencing this value, so let's not cause their apps to suddenly break with AttributeErrors (at least not yet without warning).

AttributeError is very straightforward to fix. CredentialUnavailableError (fail to connect to authorrty host) is much harder. :)

**Comment by xiangyan99** [comment]
Talked with Laurent, given the maintenance cost is low, we will keep AZURE_GERMANY to not break existing customers.

**Comment by joshfree** [comment]
> Talked with Laurent, given the maintenance cost is low, we will keep AZURE_GERMANY to not break existing customers.

This is the right decision.  We shouldn't make API breaking changes like this without following our breaking changes processes (even if using the enum value results in an error from the service).

We can talk about removing deprecated APIs like this in a major version bump in the future after we've given customers sufficient deprecated warning time and we follow our team's breaking change process

File: sdk/identity/azure-identity/azure/identity/_constants.py [hunk 1]
```diff
@@ -13,9 +14,18 @@
 CACHE_CAE_SUFFIX = ".cae"
 
 
-class AzureAuthorityHosts:
+class AzureAuthorityHostsMeta(type):
+    def __getattr__(cls, name):
+        if name == "AZURE_GERMANY":
+            warnings.warn(
+                "AZURE_GERMANY is deprecated. Please use other clouds instead.", DeprecationWarning, stacklevel=2
+            )
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/_constants.py:22]: For some wording consistency with other languages:
```suggestion
            warnings.warn(
                "AZURE_GERMANY is deprecated. Microsoft Cloud Germany was closed on October 29th, 2021.",
                DeprecationWarning,
                stacklevel=2
            )
```

**Review by pvaneck — APPROVED** [review]

**Pull request merged**